File: 0/task_scheduler.cpp
// sort() will sort the vector in ascending order as default
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
      vector<int> count(26, 0);
      for(char t : tasks)
        count[t - 'A']++;
      sort(count.begin(), count.end());
      int i = 25;
      while(i >= 0 and count[i] == count[25]) --i;
      // the 25 - i is the additional repeated chunck starter
      // see here http://www.cnblogs.com/grandyang/p/7098764.html
      return max(int(tasks.size()), (count[25] - 1) * (n + 1) + 25 - i);
    }
};

// priority queue can also work

File: 0/task_scheduler.py
# 621. Task Scheduler
# medium

from collections import defaultdict
class Solution:
    def leastInterval(self, tasks, n):
        """
        :type tasks: List[str]
        :type n: int
        :rtype: int
        """
        counter = [0] * 26
        for t in tasks:
            counter[ord(t) - ord('A')] += 1

        counter.sort()

        i = 25
        while i >= 0 and counter[25] == counter[i]:
            i -= 1

        return max(len(tasks), (counter[25] - 1) * (n + 1) + 25 - i)
